### PR TITLE
build: Bump clang minimum supported version to 14

### PR DESCRIPTION
**Pull request by maflcko**
Most supported operating systems ship with clang-14 (or later), so bump the minimum to that and allow new code to drop workarounds for previous clang bugs.

For reference:
* https://packages.debian.org/bookworm/clang (`clang-14`)
* https://packages.ubuntu.com/jammy/clang (`clang-14`)
* CentOS-like 8/9 Stream: All Clang versions from 15 to 17
* FreeBSD 12/13: All Clang versions from 15 to 16
* OpenSuse Tumbleweed ships with https://software.opensuse.org/package/clang (`clang17`); No idea about OpenSuse Leap

On operating systems where the clang version is not shipped by default, the user would have to use GCC, or install clang in a different way. For example:

* https://packages.debian.org/bullseye/g++ (g++-10)
* https://packages.ubuntu.com/focal/g++-10
* https://apt.llvm.org/, or nix, or guix, or compile clang from source, ...

### CODE REVIEW

File: src/test/fuzz/multiplication_overflow.cpp [hunk 1]
```diff
@@ -14,6 +10,14 @@
 #include <string>
 #include <vector>
 
+#if defined(__has_builtin)
+#if __has_builtin(__builtin_mul_overflow)
+#define HAVE_BUILTIN_MUL_OVERFLOW
+#endif
+#elif defined(__GNUC__)
+#define HAVE_BUILTIN_MUL_OVERFLOW
```
Comment (hebasto) [inline src/test/fuzz/multiplication_overflow.cpp:18]: nit: I understand that that was a clean commit revert. But these lines look really unneeded these days.

**Review by hebasto — APPROVED** [review]
ACK fa223ba5eb764fe822229a58d4d44d3ea83d0793.

**Review by fanquake — APPROVED** [review]
ACK aaaace2fd1299939c755c281b787df0bbf1747a0

**Review by hebasto — COMMENTED** [review]
Post-merge ACK aaaace2fd1299939c755c281b787df0bbf1747a0.

**Pull request merged**